From: Felix Fietkau <nbd@nbd.name>
Date: Wed, 3 Oct 2012 00:00:00 +0200
Subject: [PATCH] mac80211: allow scans in access point mode (for site survey)

Ported from OpenWrt's 210-ap_scan.patch (package/kernel/mac80211/patches/subsys/).
Allows AP-mode interfaces to initiate WiFi scans, which is needed for bmc-virt
where hwsim radios must scan while in AP mode to discover other virtual APs.

Without this patch, ieee80211_scan() returns -EOPNOTSUPP for AP interfaces
that are already beaconing, unless the driver advertises NL80211_FEATURE_AP_SCAN.

--- a/net/mac80211/cfg.c
+++ b/net/mac80211/cfg.c
@@ -2848,6 +2848,8 @@
 		 */
 		fallthrough;
 	case NL80211_IFTYPE_AP:
+		/* allow AP-mode scanning (site survey) */
+		break;
 		/*
 		 * If the scan has been forced (and the driver supports
 		 * forcing), don't care about being beaconing already.
